src/details/ArborX_AccessTraits.hpp

```diff
@@ -81,22 +81,25 @@ template <typename Traits>
 using AccessTraitsMemorySpaceArchetypeAlias = typename Traits::memory_space;
 
 // archetypal expression for 'size()' static member function in access traits
-template <typename Traits>
-using AccessTraitsSizeArchetypeExpression = decltype(
-    Traits::size(std::declval<first_template_parameter_t<Traits> const &>()));
+template <typename Traits, typename X>
+using AccessTraitsSizeArchetypeExpression =
+    decltype(Traits::size(std::declval<X const &>()));
 
 // archetypal expression for 'get()' static member function in access traits
-template <typename Traits>
-using AccessTraitsGetArchetypeExpression = decltype(
-    Traits::get(std::declval<first_template_parameter_t<Traits> const &>(), 0));
+template <typename Traits, typename X>
+using AccessTraitsGetArchetypeExpression =
+    decltype(Traits::get(std::declval<X const &>(), 0));
 
 template <typename Access>
-struct AccessTraitsHelper
+struct AccessTraitsHelper;
+
+template <typename X, typename Tag>
+struct AccessTraitsHelper<AccessTraits<X, Tag>>
 {
   // Deduce return type of get()
-  using type =
-      std::decay_t<detected_t<AccessTraitsGetArchetypeExpression, Access>>;
-  using tag = typename Tag<type>::type;
+  using type = std::decay_t<
+      detected_t<AccessTraitsGetArchetypeExpression, AccessTraits<X, Tag>, X>>;
+  using tag = typename ::ArborX::Details::Tag<type>::type;
```

Comment:
I'm not sure, why do you need `::ArborX::Details::` here, should i come from the `Tag` template parameter?

Reply:
Definitely need it.

Reply:
If that makes you feel any better, next PR will simplify this and remove `<ArborX_DetailsTags.hpp>` altogether.

Reply:
I'm not worried about it. I don't understand the change 100%, but I trust you are doing the right thing here. Not going to spend time thinking about it.

Reply:
Do you need me to elaborate before you approve?